storage/innobase/handler/ha_innodb.cc

```diff
@@ -20094,6 +20126,13 @@ static int innobase_xa_prepare(handlerton *hton, /*!< in: InnoDB handlerton */
 
   if (prepare_trx ||
       (!thd_test_options(thd, OPTION_NOT_AUTOCOMMIT | OPTION_BEGIN))) {
+
```

Comment:
could we add a check that trx is rw..

Reply:
I have added a top-level `assert(!trx->read_only)`. It is an aggressive assert that checks that no RO-at-InnoDB level transactions were registered as RW in the server layer, the tests pass with it.